### DISCUSSION TITLE
Can someone help me? i try to setState data from the input field, but it seem like state doesn't update immediately.

### DISCUSSION THREAD

**Discussion by YolkKanapon** [react-hook-form/react-hook-form#4382] [Q&A]
`const onSubmit = (data) => {
    console.log(data);
    setValues(data);
    console.log(values);
  }`
 
Here's an output, it will update when it triggers twice.
![Inked1615194229921_LI](https://user-images.githubusercontent.com/54444836/110300354-5b608a00-8029-11eb-8451-76cac0cad109.jpg)

**Comment by jorisre** [comment]
@yolkyboy Please provide a full repro in www.codesandbox.io thanks

**Reply by YolkKanapon** [reply]
https://codesandbox.io/s/headless-cdn-kcexq?file=/src/App.js:541-549
@jorisre

**Reply by oualidmaz** [reply]
@yolkyboy setValues is async ... that is why you are not seeing values immediately after your console.log .  setState doesn't mutate the state immediately after calling it. 
In class component, setState take a callback function as argument , so you can easily wait for the callback 
`setState(myValue, ()=>{console.log(myValue)}` 
Unfortunately useState hook doesn't have a callback function. As workaround you can use useEffect hook to wait for the state to update. 
```
 useEffect(() => {
    console.log(values);
  },[values]);
```

https://codesandbox.io/s/exciting-driscoll-zuv2v?file=/src/App.js

Note : why not consider using getValues , provided by RHF ? 
https://react-hook-form.com/api#getValues

**Reply by YolkKanapon** [reply]
Thanks, I'll try getValues.
